lib/copy/copy.js

```diff
@@ -2,51 +2,213 @@
-        ncp(src, dest, options, callback)
+  rs.on('error', err => cb(err))
+  ws.on('error', err => cb(err))
+
+  ws.on('open', () => {
+    rs.pipe(ws)
+  }).once('close', () => {
+    fs.chmod(dest, srcStat.mode, err => {
+      if (err) return cb(err)
+      if (options.preserveTimestamps) {
+        return utimes(dest, srcStat.atime, srcStat.mtime, cb)
+      } else return cb()
+    })
+  })
+}
+
+function onDir (src, srcStat, dest, options, cb) {
+  checkDest(dest, (err, res) => {
+    if (err) return cb(err)
+    if (res === DEST_NOENT) {
+      // if dest is a subdir of src, prevent copying into itself
+      if (isSrcSubdir(src, dest)) return cb(new Error(`Cannot copy directory '${src}' into itself '${dest}'`))
+      fs.mkdir(dest, srcStat.mode, err => {
+        if (err) return cb(err)
+        fs.chmod(dest, srcStat.mode, err => {
+          if (err) return cb(err)
+          return cpDir(src, dest, options, cb)
+        })
       })
+    } else if (res === DEST_EXISTS) {
+      if (isSrcSubdir(src, dest)) return cb(new Error(`Cannot copy directory '${src}' into itself '${dest}'`))
+      return cpDir(src, dest, options, cb)
+    } else if (res && typeof res !== 'number') { // dest exists and is a link
+      if (isSrcSubdir(src, res)) return cb(new Error(`Cannot copy directory '${src}' into itself '${res}'`))
+      if (src === res) return cb()
+      return cpDir(src, dest, options, cb)
+    } else return cb()
+  })
+}
+
+function cpDir (src, dest, options, cb) {
+  fs.readdir(src, (err, items) => {
+    if (err) return cb(err)
+    Promise.all(items.map(item => {
+      return new Promise((resolve, reject) => {
+        copy(path.join(src, item), path.join(dest, item), options, err => {
+          if (err) reject(err)
+          else resolve()
+        })
+      })
+    })).then(() => {
+      return cb()
+    }).catch(err => {
+      return cb(err)
+    })
+  })
+}
+
+function onLink (src, dest, options, cb) {
+  fs.readlink(src, (err, resolvedSrcPath) => {
+    if (err) return cb(err)
+
+    if (options.dereference) {
+      resolvedSrcPath = path.resolve(process.cwd(), resolvedSrcPath)
+    }
+
+    checkDest(dest, (err, resolvedDestPath) => {
+      if (err) return cb(err)
+
+      if (resolvedDestPath === DEST_NOENT) {
+        // if dest is a subdir of resolved src path, prevent copying into itself
+        if (isSrcSubdir(resolvedSrcPath, dest)) {
+          return cb(new Error(`Cannot copy directory '${resolvedSrcPath}' into itself '${dest}'`))
+        }
+        return fs.symlink(resolvedSrcPath, dest, cb)
+      } else if (resolvedDestPath === DEST_EXISTS) { // dest exists but is not a link
+        if (isSrcSubdir(resolvedSrcPath, dest)) {
+          return cb(new Error(`Cannot copy directory '${resolvedSrcPath}' into itself '${dest}'`))
+        }
+        // if src points to dest
+        if (resolvedSrcPath === dest) return cb()
+        return copy(resolvedSrcPath, dest, options, cb)
```

Comment:
@RyanZim I guess you are right. I don't have a good feeling about `return copy(resolvedSrcPath, dest, options, cb)` neither. I recall I used it because of the ability to handle weird cases like when src is a symlink and dest exists (either regular or another symlink) and is a kid of src as well.

`return fs.symlink(resolvedSrcPath, dest, cb)` correctly throws `EEXIST: file already exists, symlink 'src' -> 'dest'` for those cases.

Now, I think we should decide what we want to do here for these two cases when src is a symlink:

- dest exists and is a regular file/directory
- dest exists and is a symlink (file/directory)

Do we want to throw an error and notify user that dest exists?